[Event Grid] Add Control Plane API Version 2025-04-01-preview

bbi2015 opened the pull request:
# ARM (Control Plane) API Specification Update Pull Request 

> [!TIP]
> Overwhelmed by all this guidance? See the `Getting help` section at the bottom of this PR description.

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

![spec_pr_review_workflow_diagram](https://github.com/Azure/azure-rest-api-specs/assets/4429827/5bb5e7ce-8aff-4dbb-a3f8-0d9b68fef5b1)

## Purpose of this PR

What's the purpose of this PR? Check the specific option that applies. This is **mandatory**!

  - [ ] New resource provider.
  - [x] New API version for an existing resource provider. (If API spec is not defined in TypeSpec, the PR should have been created in adherence to [OpenAPI specs PR creation guidance](https://aka.ms/azsdkdocs/createopenapispec)).
  - [ ] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix OpenAPI spec quality issues in S360.
  - [ ] Convert existing [OpenAPI spec to TypeSpec spec](https://aka.ms/typespec/conversion) (do not combine this with implementing changes for a new API version).
  - [ ] Other, please clarify:
    - _edit this with your clarification_

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [x] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data plane related specifications.
- [x] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://aka.ms/azurerpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can proceed to the diagram Step 2, "ARM API changes review", for this PR.

## Additional information

<details>
<summary> Viewing API changes</summary>

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

</details>
<details>
<summary>Suppressing failures</summary>

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[suppressions guide](https://aka.ms/azsdk/pr-suppressions) to get approval.

</details>

## Getting help

- First, please carefully read through this PR description, from top to bottom. Please fill out the `Purpose of this PR` and `Due diligence checklist`.
- If you don't have permissions to remove or add labels to the PR, request `write access` per [aka.ms/azsdk/access#request-access-to-rest-api-or-sdk-repositories](https://aka.ms/azsdk/access#request-access-to-rest-api-or-sdk-repositories)
- To understand what you must do next to merge this PR, see the `Next Steps to Merge` comment. It will appear within few minutes of submitting this PR and will continue to be up-to-date with current PR state.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- For help with ARM review (PR workflow diagram Step 2), see https://aka.ms/azsdk/pr-arm-review.
- If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
  This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

bbi2015 committed:
Add previous 2024-12-15-preview version files as baseline

bbi2015 committed:
Add 2025-04-15-preview API version files

bbi2015 committed:
Fix API version in response header of example files

razvanbadea-msft commented:
>         "userAssignedIdentity": {

maybe reference the UserAssignedIdentity from common-types?

---
Refers to: specification/eventgrid/resource-manager/Microsoft.EventGrid/preview/2025-04-15-preview/EventGrid.json:17813 in 9c8cf22. [](commit_id = 9c8cf22fac10e8bbe327a3fb1ca79708ffc714fb, deletion_comment = False)

razvanbadea-msft commented:
>           "description": "Azure Active Directory Application ID or URI to get the access token that will be included as the bearer token in delivery requests.",

use the new name for AAD - Microsoft Entra Id in descriptions for clarity 

---
Refers to: specification/eventgrid/resource-manager/Microsoft.EventGrid/preview/2025-04-15-preview/EventGrid.json:17842 in 9c8cf22. [](commit_id = 9c8cf22fac10e8bbe327a3fb1ca79708ffc714fb, deletion_comment = False)

razvanbadea-msft commented:
>           "type": "string"

you can add uri format for these properties

---
Refers to: specification/eventgrid/resource-manager/Microsoft.EventGrid/preview/2025-04-15-preview/EventGrid.json:17839 in 9c8cf22. [](commit_id = 9c8cf22fac10e8bbe327a3fb1ca79708ffc714fb, deletion_comment = False)

razvanbadea-msft commented:
Fix the required checks which fails

bbi2015 committed:
Add suppressions to fix Swagger LintDiff

a-hamad commented:
> > ```
> >     "userAssignedIdentity": {
> > ```
> 
> maybe reference the UserAssignedIdentity from common-types?
> 
> Refers to: specification/eventgrid/resource-manager/Microsoft.EventGrid/preview/2025-04-15-preview/EventGrid.json:17813 in [9c8cf22](https://github.com/Azure/azure-rest-api-specs/commit/9c8cf22fac10e8bbe327a3fb1ca79708ffc714fb). [](commit_id = [9c8cf22](https://github.com/Azure/azure-rest-api-specs/commit/9c8cf22fac10e8bbe327a3fb1ca79708ffc714fb), deletion_comment = False)

Thanks for the feedback. We would like to keep this as is for now to maintain consistency in the swagger file where similar property is used.

We will consider updating the whole swagger later with this change if this is really needed. Having two ways to reference 'userAssignedIdentity' is something we want to avoid.

a-hamad commented:
> 17839

Thanks for your feedback. We have other properties in swagger which represent base URL and we been using type: string for them.

Example:

Line 11971 --
 "endpointBaseUrl": {
          "description": "The base URL that represents the endpoint of the partner destination.",
          "type": "string"
        },

Line 14054 --
      "endpointBaseUrl": {
          "description": "The base URL that represents the endpoint of the destination of an event subscription.",
          "type": "string",
          "readOnly": true
        },

and others.

for sake of consistency, we would like to keep this one the same way.

a-hamad commented:
> Thanks for your feedback. We have other properties in swagger which represent base URL and we been using type: string for them.
> 
> Example:
> 
> Line 11971 -- "endpointBaseUrl": { "description": "The base URL that represents the endpoint of the partner destination.", "type": "string" },
> 
> Line 14054 -- "endpointBaseUrl": { "description": "The base URL that represents the endpoint of the destination of an event subscription.", "type": "string", "readOnly": true },
> 
> and others.
> 
> for sake of consistency, we would like to keep this one the same way.


Thanks for your feedback. We have other properties in swagger which represent base URL and we been using type: string for them.

Example:

Line 11971 --
"endpointBaseUrl": {
"description": "The base URL that represents the endpoint of the partner destination.",
"type": "string"
},

Line 14054 --
"endpointBaseUrl": {
"description": "The base URL that represents the endpoint of the destination of an event subscription.",
"type": "string",
"readOnly": true
},

and others.

for sake of consistency, we would like to keep this one the same way.

a-hamad commented:
> Fix the required checks which fails

For the Avocado failures, https://github.com/Azure/azure-rest-api-specs/pull/33889/checks?check_run_id=40403726726

These seem to be infra/false-positive failures. These are already deprecated APIs and/or resources that are not used in previous public preview/stable swagger.

How can we address/suppress these failures?

bbi2015 committed:
Add suppressions for LintDiff and fix to rename Azure Active Directory to Microsoft Entra ID

bbi2015 committed:
Fix API Path in suppressions rules

bbi2015 committed:
Fixing property path for LintDiff error

bbi2015 committed:
Fix LintDiff issues

bbi2015 committed:
Fix LintDiff warnings regarding use of latest version of common types

bbi2015 committed:
Fix use of older ErrorResponse schema

bbi2015 committed:
Rename all renaming AAD to Microsoft Entra ID

bbi2015 commented:
> > ```
> >       "description": "Azure Active Directory Application ID or URI to get the access token that will be included as the bearer token in delivery requests.",
> > ```
> 
> use the new name for AAD - Microsoft Entra Id in descriptions for clarity
> 
> Refers to: specification/eventgrid/resource-manager/Microsoft.EventGrid/preview/2025-04-15-preview/EventGrid.json:17842 in [9c8cf22](https://github.com/Azure/azure-rest-api-specs/commit/9c8cf22fac10e8bbe327a3fb1ca79708ffc714fb). [](commit_id = [9c8cf22](https://github.com/Azure/azure-rest-api-specs/commit/9c8cf22fac10e8bbe327a3fb1ca79708ffc714fb), deletion_comment = False)

Thanks for the feedback. I have renamed AAD to Microsoft Entra ID in all the places.

bbi2015 commented:
> @bbi2015 please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

@microsoft-github-policy-service agree company="Microsoft"

bbi2015 commented:
There are 3 [Swagger Avocado Failures](https://github.com/Azure/azure-rest-api-specs/pull/33889/checks?check_run_id=40574012490), which seems like false positive. All the errors are related to "MISSING_APIS_IN_DEFAULT_TAG". Please find the analysis below.

In Summary, the errors are related to paths which are no longer supported and have been removed in earlier preview versions. 

   1.  The API path "/subscriptions/{}/resourcegroups/{}/providers/microsoft.eventgrid/partnernamespaces/{}/eventchannels/{}" has been deprecated as part of API version 2023-06-01-preview. The API path is last supported in API version 2021-10-15-preview.
        - Related PR : https://github.com/Azure/azure-rest-api-specs/pull/23981
   2. The API path "/subscriptions/{}/resourcegroups/{}/providers/microsoft.eventgrid/partnernamespaces/{}/eventchannels" has been deprecated as part of API version 2023-06-01-preview. The API path is last supported in  API version 2021-10-15-preview.
        - Related PR : https://github.com/Azure/azure-rest-api-specs/pull/23981
   3. The API path "/providers/microsoft.eventgrid/partnerregistrations" has been deprecated as part of API version 2021-06-01-preview. The API path is last supported in API version 2020-10-15-preview.
        - Related PR : https://github.com/Azure/azure-rest-api-specs/pull/14783.

gary-x-li commented:
> > Thanks for your feedback. We have other properties in swagger which represent base URL and we been using type: string for them.
> > Example:
> > Line 11971 -- "endpointBaseUrl": { "description": "The base URL that represents the endpoint of the partner destination.", "type": "string" },
> > Line 14054 -- "endpointBaseUrl": { "description": "The base URL that represents the endpoint of the destination of an event subscription.", "type": "string", "readOnly": true },
> > and others.
> > for sake of consistency, we would like to keep this one the same way.
> 
> Thanks for your feedback. We have other properties in swagger which represent base URL and we been using type: string for them.
> 
> Example:
> 
> Line 11971 -- "endpointBaseUrl": { "description": "The base URL that represents the endpoint of the partner destination.", "type": "string" },
> 
> Line 14054 -- "endpointBaseUrl": { "description": "The base URL that represents the endpoint of the destination of an event subscription.", "type": "string", "readOnly": true },
> 
> and others.
> 
> for sake of consistency, we would like to keep this one the same way.

You can keep `type: string` and add `format: uri` to indicate it's a URI. Will you be able to apply this to all similar properties?

gary-x-li commented:
> > > ```
> > >     "userAssignedIdentity": {
> > > ```
> > 
> > 
> > maybe reference the UserAssignedIdentity from common-types?
> > Refers to: specification/eventgrid/resource-manager/Microsoft.EventGrid/preview/2025-04-15-preview/EventGrid.json:17813 in [9c8cf22](https://github.com/Azure/azure-rest-api-specs/commit/9c8cf22fac10e8bbe327a3fb1ca79708ffc714fb). [](commit_id = [9c8cf22](https://github.com/Azure/azure-rest-api-specs/commit/9c8cf22fac10e8bbe327a3fb1ca79708ffc714fb), deletion_comment = False)
> 
> Thanks for the feedback. We would like to keep this as is for now to maintain consistency in the swagger file where similar property is used.
> 
> We will consider updating the whole swagger later with this change if this is really needed. Having two ways to reference 'userAssignedIdentity' is something we want to avoid.

Is this the user assigned identity resource ID such as `/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ManagedIdentity/userAssignedIdentities/{identityName}`? If so, consider adding `"format": "arm-id"` and [x-ms-arm-id-details](https://azure.github.io/autorest/extensions/#x-ms-arm-id-details) to limit to identity resource type.

bbi2015 committed:
Add format to endpointUrl and endpointBaseUrl and x-ms-arm-od-details to userAssignedIdentity property

bbi2015 committed:
Merge remote tracking branch 'upstream/main' to run main gates check

bbi2015 committed:
Resolving merge conflicts

bbi2015 committed:
Removing x-ms-arm-id-details from existing properties

bbi2015 commented:
> > > Thanks for your feedback. We have other properties in swagger which represent base URL and we been using type: string for them.
> > > Example:
> > > Line 11971 -- "endpointBaseUrl": { "description": "The base URL that represents the endpoint of the partner destination.", "type": "string" },
> > > Line 14054 -- "endpointBaseUrl": { "description": "The base URL that represents the endpoint of the destination of an event subscription.", "type": "string", "readOnly": true },
> > > and others.
> > > for sake of consistency, we would like to keep this one the same way.
> > 
> > 
> > Thanks for your feedback. We have other properties in swagger which represent base URL and we been using type: string for them.
> > Example:
> > Line 11971 -- "endpointBaseUrl": { "description": "The base URL that represents the endpoint of the partner destination.", "type": "string" },
> > Line 14054 -- "endpointBaseUrl": { "description": "The base URL that represents the endpoint of the destination of an event subscription.", "type": "string", "readOnly": true },
> > and others.
> > for sake of consistency, we would like to keep this one the same way.
> 
> You can keep `type: string` and add `format: uri` to indicate it's a URI. Will you be able to apply this to all similar properties?

Hi @gary-x-li ,

I have updated and PR and added `format: uri` to all new properties. I have added `format: uri` to old properties as well, but seems like this fails the Breaking Change Test [[Breaking Change Link](https://github.com/Azure/azure-rest-api-specs/runs/40956322712)]. Hence for now we would like to keep the existing ones as it is and incorporated your suggestions in the newly added properties. 

The same behavior has been observed when I have added "format": "arm-id" and [x-ms-arm-id-details](https://azure.github.io/autorest/extensions/#x-ms-arm-id-details) to limit to identity resource type.

bbi2015 committed:
Renaming API version from 2025-04-15-preview to 2025-04-01-preview

a-hamad commented on specification/eventgrid/resource-manager/readme.ruby.md:
````diff
@@ -33,6 +34,16 @@ batch:
+
+``` yaml $(tag) == 'package-2025-04-preview' && $(ruby)
+
+namespace: "Azure::EventGrid::Mgmt::V2025_04_15_preview"
````
V2025_04_15_preview --> V2025_04_01_preview

bbi2015 committed:
Fixing readme.ruby.md with current api version

a-hamad reviewed (approved):
LGTM

bbi2015 committed:
Merge remote-tracking branch 'upstream/main' into users/adebnath/eventgrid

bbi2015 committed:
Merge remote-tracking branch 'upstream/main' into users/adebnath/eventgrid
